lib/contracts.rb

```diff
@@ -242,108 +267,110 @@ def call(*args, &blk)
+  # it's possible we are going to pass in a block instead. So lets
+  # append a nil to the list of args just so it doesn't fail.
 
-    args_size = args.size
-    _args_size = _args.size
-    contracts_size = args_contracts.size
+  # a better way to handle this might be to take this into account
+  # before throwing a "mismatched # of args" error.
+  def maybe_append_block! args, blk
+    if has_proc_contract? && !blk &&
+      (args_contract_index || args.size < args_contracts.size)
+      args << nil
+    end
+  end
 
-    # Explicitly append blk=nil if nil != Proc contract violation anticipated
-    if @has_proc_contract && !blk &&
-      (@splat_lower_index || _args_size < contracts_size)
-      _args << nil
-      _args_size += 1
+  # Same thing for when we have named params but didn't pass any in.
+  def maybe_append_options! args, blk
+    return unless has_options_contract?
+    if has_proc_contract? && Hash === args_contracts[-2] && !args[-2].is_a?(Hash)
+      args.insert(-2, {})
+    elsif Hash === args_contracts[-1] && !args[-1].is_a?(Hash)
+      args << {}
     end
+  end
 
-    # Explicitly append options={} if Hash contract is present
-    if @has_options_contract
-      if @has_proc_contract && Hash === @args_contracts[-2] && !_args[-2].is_a?(Hash)
-        _args.insert(-2, {})
-        _args_size += 1
-      elsif Hash === @args_contracts[-1] && !_args[-1].is_a?(Hash)
-        _args << {}
-        _args_size += 1
-      end
+  def validate_arg(arg, contract, validator, arg_pos, args)
+    if validator && validator[arg]
+      true
+    else
+      Contract.failure_callback({
+        :arg => arg,
+        :contract => contract,
+        :class => klass,
+        :method => method,
+        :contracts => self,
+        :arg_pos => arg_pos,
+        :total_args => args.size
+      })
     end
+  end
 
-    # Loop forward validating the arguments up to the splat (if there is one)
-    (@splat_lower_index || _args_size).times do |i|
-      _arg = _args[i]
-      contract = @args_contracts[i]
-      validator = @args_validators[i]
-
-
-      unless validator && validator[_arg]
-        return unless Contract.failure_callback({
-          :arg => _arg,
-          :contract => contract,
-          :class => @klass,
-          :method => @method,
-          :contracts => self,
-          :arg_pos => i + 1,
-          :total_args => args_size
-        })
-      end
+  def call_with(this, *args, &blk)
+    args << blk if blk
+
+    # Explicitly append blk=nil if nil != Proc contract violation anticipated
+    maybe_append_block!(args, blk)
+
+    # Explicitly append options={} if Hash contract is present
+    maybe_append_options!(args, blk)
 
+    # Loop forward validating the arguments up to the splat (if there is one)
+    (args_contract_index || args.size).times do |i|
+      contract = args_contracts[i]
+      arg = args[i]
+      validator = args_validators[i]
+      return unless validate_arg(arg, contract, validator, i+1, args)
```

Comment:
This reads much better, but as your blog post shows method calls are slow, calling `validate_arg` in a loop may be unacceptable. Benchmark, inline it then bench again to see if it's a bottle neck, that's my hunch. I wonder if a lambda local variable would be faster than a full method here as an alternative.